compile.sh

```diff
@@ -518,53 +638,99 @@ buildtool="" #Use ninja by default, fall back to make if that is not available.
+      if [ -e data/locale ] ; then
+        # Prefer link to default builddir
+        if using_default_builddir ; then
+          $RUN rm data/locale
+        else
+          return 0
+        fi
+      fi
+      $RUN ln -s ../"$BUILDDIR"/locale data/locale
+    fi
     return 0
   }
 
-  # Compile Widelands
+### Function to actually compile Widelands
   compile_widelands () {
-    cmake $GENERATOR .. $EXTRA_OPTS -DCMAKE_BUILD_TYPE=$BUILD_TYPE -DOPTION_BUILD_WEBSITE_TOOLS=$BUILD_WEBSITE -DOPTION_BUILD_TRANSLATIONS=$BUILD_TRANSLATIONS -DOPTION_BUILD_TESTS=$BUILD_TESTS -DOPTION_ASAN=$USE_ASAN -DOPTION_TSAN=$USE_TSAN -DUSE_XDG=$USE_XDG -DUSE_FLTO_IF_AVAILABLE=${USE_FLTO}
-
-    $buildtool -j $CORES
+    $RUN cmake $GENERATOR  "$SOURCEDIR"  $EXTRA_OPTS           \
+               -DCMAKE_BUILD_TYPE=$BUILD_TYPE                  \
+               -DOPTION_BUILD_WEBSITE_TOOLS=$BUILD_WEBSITE     \
+               -DOPTION_BUILD_TRANSLATIONS=$BUILD_TRANSLATIONS \
+               -DOPTION_BUILD_TESTS=$BUILD_TESTS               \
+               -DOPTION_ASAN=$USE_ASAN                         \
+               -DOPTION_TSAN=$USE_TSAN                         \
+               -DUSE_XDG=$USE_XDG                              \
+               -DUSE_FLTO_IF_AVAILABLE=${USE_FLTO}
+
+    $RUN $buildtool -j $CORES
 
     return 0
   }
 
-  # Remove old and move newly compiled files
+### Function to remove old and move newly compiled files
   move_built_files () {
-    rm  -f ../VERSION || true
-    rm  -f ../widelands || true
+    # Only replace files for the default build, except for main executable,
+    # which is also replaced in case of -e or --exename
+    if ! using_default_builddir ; then
+      if [ -n "$EXENAME" ]; then
+        # This is called in $BUILDDIR
+        $RUN rm -f "${SOURCEDIR}/$EXENAME" || true
+        $RUN mv src/widelands "${SOURCEDIR}/$EXENAME"
+        return 0
+      fi
+      return 1
+    fi
 
-    rm  -f ../wl_map_object_info || true
-    rm  -f ../wl_map_info || true
+    if [ -z "$EXENAME" ]; then
+      EXENAME=widelands
+    fi
 
-    cp VERSION ../VERSION
-    mv src/widelands ../widelands
+    # This is called in $BUILDDIR, but only if it is ${SOURCEDIR}/$BUILDDIR_DEFAULT,
+    # so .. is $SOURCEDIR
+    $RUN rm  -f ../VERSION || true
+    $RUN rm  -f ../"$EXENAME" || true
+
+    $RUN rm  -f ../wl_map_object_info || true
+    $RUN rm  -f ../wl_map_info || true
+
+    $RUN cp VERSION ../VERSION
+    $RUN mv src/widelands ../"$EXENAME"
 
     if [ $BUILD_WEBSITE = "ON" ]; then
-        mv ../build/src/website/wl_create_spritesheet ../wl_create_spritesheet
-        mv ../build/src/website/wl_map_object_info ../wl_map_object_info
-        mv ../build/src/website/wl_map_info ../wl_map_info
+      $RUN mv ../build/src/website/wl_create_spritesheet ../wl_create_spritesheet
+      $RUN mv ../build/src/website/wl_map_object_info ../wl_map_object_info
+      $RUN mv ../build/src/website/wl_map_info ../wl_map_info
```

Comment:
If you include $BUILDDIR in the path, these binaries can still be moved to the source dir.